File: terceiro_semestre/tecnicas/prova_a2/python/olimpiadas.py
import sys
from collections import defaultdict

def principal():
    medalhas = defaultdict(lambda: [0, 0, 0])

    try:
        while True:
            prova = input().strip()
            ouro = input().strip()
            prata = input().strip()
            bronze = input().strip()

            medalhas[ouro][0] += 1
            medalhas[prata][1] += 1
            medalhas[bronze][2] += 1
    except EOFError:
        pass

    medalhas_ordenadas = sorted(medalhas.items(), key=lambda x: (-x[1][0], -x[1][1], -x[1][2], x[0]))

    print("Quadro de Medalhas")
    for pais, (ouro, prata, bronze) in medalhas_ordenadas:
        print(f"{pais} {ouro} {prata} {bronze}")

if __name__ == "__main__":
    principal()

File: terceiro_semestre/tecnicas/prova_a2/c++/olimpiadas.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

struct Medalhas {
    int ouro = 0;
    int prata = 0;
    int bronze = 0;
};

int main() {
    unordered_map<string, Medalhas> quadro;
    string modalidade, ouro, prata, bronze;

    while (getline(cin, modalidade)) {
        getline(cin, ouro);
        getline(cin, prata);
        getline(cin, bronze);

        quadro[ouro].ouro++;
        quadro[prata].prata++;
        quadro[bronze].bronze++;
    }

    vector<pair<string, Medalhas>> paises(quadro.begin(), quadro.end());

    sort(paises.begin(), paises.end(), [](const pair<string, Medalhas>& a, const pair<string, Medalhas>& b) {
        if (a.second.ouro != b.second.ouro) {
            return a.second.ouro > b.second.ouro;
        }
        if (a.second.prata != b.second.prata) {
            return a.second.prata > b.second.prata;
        }
        if (a.second.bronze != b.second.bronze) {
            return a.second.bronze > b.second.bronze;
        }
        return a.first < b.first;
    });

    cout << "Quadro de Medalhas" << "\n";
    for (const auto& pais : paises) {
        cout << pais.first << " " << pais.second.ouro << " " << pais.second.prata << " " << pais.second.bronze << "\n";
    }

    return 0;
}
